src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java

```diff
@@ -16,317 +16,507 @@
+    }
+    curColIdx++;
   }
 
   private int deserializeValidityBuffer(ColumnOffsetInfo curColOffset) {
     if (curColOffset.getValidity() != INVALID_OFFSET) {
-      long offset = curColOffset.getValidity();
-      long validityBufferSize = padFor64byteAlignment(getValidityLengthInBytes(getTotalRowCount()));
-      try (HostMemoryBuffer validityBuffer = buffer.slice(offset, validityBufferSize)) {
-        int nullCountTotal = 0;
-        int startRow = 0;
-        for (int tableIdx = 0; tableIdx < getTableSize(); tableIdx += 1) {
-          SliceInfo sliceInfo = sliceInfoOf(tableIdx);
-          long validityOffset = validifyBufferOffset(tableIdx);
-          if (validityOffset != INVALID_OFFSET) {
-            nullCountTotal += copyValidityBuffer(validityBuffer, startRow,
-                memoryBufferOf(tableIdx), toIntExact(validityOffset),
-                sliceInfo);
-          } else {
-            appendAllValid(validityBuffer, startRow, sliceInfo.getRowCount());
-          }
-
-          startRow += sliceInfo.getRowCount();
+      int offset = toIntExact(curColOffset.getValidity());
+      int nullCountTotal = 0;
+      int startRow = 0;
+      for (int tableIdx = 0; tableIdx < kudoTables.length; tableIdx += 1) {
+        SliceInfo sliceInfo = sliceInfoOf(tableIdx);
+        long validityOffset = validityOffsets[tableIdx];
+        if (kudoTables[tableIdx].getHeader().hasValidityBuffer(curColIdx)) {
+          nullCountTotal += copyValidityBuffer(buffer, offset, startRow,
+                  kudoTables[tableIdx].getBuffer(),
+                  toIntExact(validityOffset),
+                  sliceInfo, inputBuf, outputBuf);
+        } else {
+          appendAllValid(buffer, offset, startRow, sliceInfo.getRowCount());
         }
-        return nullCountTotal;
+
+        startRow += sliceInfo.getRowCount();
       }
+      return nullCountTotal;
     } else {
       return 0;
     }
   }
 
   /**
    * Copy a sliced validity buffer to the destination buffer, starting at the given bit offset.
+   * Visible for testing.
    *
    * @return Number of nulls in the validity buffer.
    */
-  private static int copyValidityBuffer(HostMemoryBuffer dest, int startBit,
-                                        HostMemoryBuffer src, int srcOffset,
-                                        SliceInfo sliceInfo) {
-    int nullCount = 0;
-    int totalRowCount = sliceInfo.getRowCount();
-    int curIdx = 0;
-    int curSrcByteIdx = srcOffset;
-    int curSrcBitIdx = sliceInfo.getValidityBufferInfo().getBeginBit();
-    int curDestByteIdx = startBit / 8;
-    int curDestBitIdx = startBit % 8;
-
-    while (curIdx < totalRowCount) {
-      int leftRowCount = totalRowCount - curIdx;
-      int appendCount;
-      if (curDestBitIdx == 0) {
-        appendCount = min(8, leftRowCount);
-      } else {
-        appendCount = min(8 - curDestBitIdx, leftRowCount);
-      }
+  static int copyValidityBuffer(HostMemoryBuffer dest, int startOffset, int startBit,
```

Comment:
nit: This method is way too long. Can you please try to break it up into smaller pieces so it is simpler to read and understand?